### PR TITLE
Only authenticate requests for private data, move away from UUIDs to identify users

### PR DESCRIPTION
**Pull request by pmarsh-scottlogic**
Started to make tests pass

In this PR, I've done two major things:
- Modified all the endpoints to all include /public or /private. The public ones have no authentication on them, whereas the private ones require a JWT.
- I've realised that we're using usernames in the persistence layer to uniquely identify users. Therefore it's no use using the accountId (uuid) to identify them too. Thus anytime we want to uniquely refer to a user, we now use their username to do this.

I'm sure I did more things that I've forgotten - I was not very well disciplined with my commit messages (sorry)!

### CODE REVIEW

File: src/main/java/com/example/demo/matcher/InitialiseFakeMarket.java [hunk 1]
```diff
@@ -12,7 +12,7 @@ public static Order fakeOrder() {
         Random rd = new Random();
         boolean bool = rd.nextBoolean();
         return new Order(
-                UUID.randomUUID(),
+                UUID.randomUUID().toString(), // These should be usernames, but usernames should be unique, so we're using uuids as usernames
```
Comment (nataliadombrowska) [inline src/main/java/com/example/demo/matcher/InitialiseFakeMarket.java:15]: Sounds good to me 👍

File: src/main/java/com/example/demo/matcher/MatcherController.java [hunk 2]
```diff
@@ -27,57 +26,55 @@ public class MatcherController {
...
-            return ResponseEntity.ok(orderService.getOrderbook(OrderAction.BUY, UUID.fromString(accountId)));
+            return ResponseEntity.ok(orderService.getOrderbook(OrderAction.BUY, username));
         }
         catch(Exception e) {
             return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
         }
     }
 
-    @GetMapping(value = "/orderbook/sell")
+    @GetMapping(value = "/public/orderbook/sell")
```
Comment (nataliadombrowska) [inline src/main/java/com/example/demo/matcher/MatcherController.java:44]: So this one won't be authenticated? Is that why you're specifying it's public?

File: src/main/java/com/example/demo/matcher/models/NewOrderParams.java [hunk 3]
```diff
@@ -12,8 +11,7 @@
 
 @ToString @AllArgsConstructor @Getter
 public class NewOrderParams {
-    @UUID(message = "Bad UUID format")
-    String account;
+    String username;
```
Comment (nataliadombrowska) [inline src/main/java/com/example/demo/matcher/models/NewOrderParams.java:14]: Are we not having any restrictions when it comes to usernames? `@Size` could be something you add here so we're still doing some sort of validation. Completely up to you though! Not a blocker just something to consider 👍

File: src/main/java/com/example/demo/security/configuration/ApplicationSecurity.java [hunk 4]
```diff
@@ -52,11 +52,12 @@ protected void configure(HttpSecurity http) throws Exception {
 
         // authorises any request via /auth/login to enter without authentication
         http.authorizeRequests()
-                .antMatchers("/auth/login").permitAll()
+                .antMatchers("/auth/login", "/h2-console/**", "/public/**").permitAll()
                 .anyRequest().authenticated();
 
         // This exception handling code ensures that the server will return HTTP status 401 (Unauthorized)
         // if any error occurs during authentication process
+        // todo: this is not what I want. It is causing nonsense urls to return 401 unauthorised errors. Which is not useful
```
Comment (nataliadombrowska) [inline src/main/java/com/example/demo/security/configuration/ApplicationSecurity.java:60]: Did you follow a tutorial for the exception handling is that why it's behaving this way?

**Review by nataliadombrowska — APPROVED** [review]
Nice good job!

File: src/main/java/com/example/demo/matcher/MatcherController.java (hunk 2 above)
Comment (pmarsh-scottlogic) [reply src/main/java/com/example/demo/matcher/MatcherController.java:44]: Yup
That way it's clear to the person programming the client that they don't need a token.
But also it makes it easy for me to configure which endpoints are authenticated and which aren't!

File: src/main/java/com/example/demo/matcher/models/NewOrderParams.java (hunk 3 above)
Comment (pmarsh-scottlogic) [reply src/main/java/com/example/demo/matcher/models/NewOrderParams.java:14]: This is something I will think about. I've made an issue already https://github.com/pmarsh-scottlogic/JavaTraining/issues/17

File: src/main/java/com/example/demo/security/configuration/ApplicationSecurity.java (hunk 4 above)
Comment (pmarsh-scottlogic) [reply src/main/java/com/example/demo/security/configuration/ApplicationSecurity.java:60]: yeah..
Without this bit of code, we get a 403 response for trying to access a private endpoint without a token

Comment (nataliadombrowska) [reply src/main/java/com/example/demo/security/configuration/ApplicationSecurity.java:60]: Let me know if you need me to look at anything!

**Pull request merged**